Frontend_Web_Development_React_Angular_Vue/React/Week 2/Functionalandclasscomponents.md

```diff
@@ -0,0 +1,228 @@
+
+This component can be tricky to change because of all the nesting, and it is also hard to reuse individual parts of it. Let’s extract a few components from it.
+
+*First, we will extract Avatar:*
+
+```javascript
+function Avatar(props) {
+  return (
+    <img className="Avatar"
+      src={props.user.avatarUrl}
+      alt={props.user.name}
+    />
+  );
+}
+```
+The **Avatar** doesn’t need to know that it is being rendered inside a Comment. This is why we have given its prop a more generic name: user rather than author.
+
+We recommend naming props from the component’s own point of view rather than the context in which it is being used.
+
+*We can now simplify Comment a tiny bit:*
+
+```javascript
+function Comment(props) {
+  return (
+    <div className="Comment">
+      <div className="UserInfo">
+        <Avatar user={props.author} />
+        <div className="UserInfo-name">
+          {props.author.name}
+        </div>
+      </div>
+      <div className="Comment-text">
+        {props.text}
+      </div>
+      <div className="Comment-date">
+        {formatDate(props.date)}
+      </div>
+    </div>
+  );
+}
+```
+*Next, we will extract a UserInfo component that renders an Avatar next to the user’s name:*
+```javascript
+function UserInfo(props) {
+  return (
+    <div className="UserInfo">
+      <Avatar user={props.user} />
+      <div className="UserInfo-name">
+        {props.user.name}
+      </div>
+    </div>
+  );
+  ```
+  *This lets us simplify Comment even further:*
+```javascript
+function Comment(props) {
+  return (
+    <div className="Comment">
+      <UserInfo user={props.author} />
+      <div className="Comment-text">
+        {props.text}
+      </div>
+      <div className="Comment-date">
+        {formatDate(props.date)}
+      </div>
+    </div>
+  );
+}
+```
+Extracting components might seem like grunt work at first, but having a palette of reusable components pays off in larger apps. A good rule of thumb is that if a part of your UI is used several times **(Button, Panel, Avatar)**, or is complex enough on its own **(App, FeedStory, Comment)**, it is a good candidate to be extracted to a separate component.
+<hr>
+
+### **Props are Read-Only**
+
+Whether you declare a component as a function or a class, it must never modify its own props. Consider this sum function:
+```javascript
+function sum(a, b) {
+  return a + b;
+}
+```
+Such functions are called “pure” because they do not attempt to change their inputs, and always return the same result for the same inputs.
+
+*In contrast, this function is impure because it changes its own input:*
+```javascript
+function withdraw(account, amount) {
+  account.total -= amount;
+}
+```
+
+*React is pretty flexible but it has a single strict rule:*
+
+**All React components must act like pure functions with respect to their props.**
+
```

Comment:
Try to add reference link at the bottom from where you have referred and created this PR.